### ISSUE TITLE
How to add the Underline plugin to the ClassicEditor build in the Angular integration?

### ISSUE THREAD

**Issue by Eyegabry99** [ckeditor/ckeditor5-angular#156]
## 📝 Provide detailed reproduction steps (if any)

1. I downloaded the classic build of CKEditor 5 and integrated it into my angular application
2. There isn't any underline plugin so i decided to install it following this guide https://ckeditor.com/docs/ckeditor5/latest/builds/guides/integration/installing-plugins.html
3. Did all of this steps and run my application

### ✔️ Expected result

*What is the expected result of the above steps?*
A working editor with underline option

### ❌ Actual result

*What is the actual result of the above steps?*
An editor without the underline option

## 📃 Other details

* Browser: Chrome 
* OS: Windows 10
* CKEditor version: 5
* Installed CKEditor plugins: Underline
---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

**Comment by ma2ciek** [comment]
Hi @Eyegabry99!

I transferred your issue.

Please, show the editor configuration you used for the `<ckeditor>` component.

**Comment by Eyegabry99** [comment]
Ok here are some screenshots. 
![image](https://user-images.githubusercontent.com/28194225/68482881-c5e99600-023a-11ea-8150-73badf1eff96.png)
![image](https://user-images.githubusercontent.com/28194225/68482903-ce41d100-023a-11ea-81df-7da6e51706ef.png)
![image](https://user-images.githubusercontent.com/28194225/68482926-d994fc80-023a-11ea-8de2-e48ee931bd83.png)
![image](https://user-images.githubusercontent.com/28194225/68482941-e285ce00-023a-11ea-8494-8beba6b64aaa.png)
And here i added the underline plugin
![image](https://user-images.githubusercontent.com/28194225/68482992-f9c4bb80-023a-11ea-9665-4d1487a082db.png)

**Comment by ma2ciek** [comment]
Hi @Eyegabry99,

Please, don't use screenshots in the bug reports as they aren't useful for anybody trying to help you. Could you change your post and paste the code instead?

**Comment by Eyegabry99** [comment]
Yeah sorry 
**component.ts**
Import of my ckeditor
`import * as ClassicEditor from '@ckeditor/ckeditor5-build-classic';`
**Editor configuration**
```
public Editor = ClassicEditor;
  optionsNews = {
    toolbar: [
      'bold', 'italic', 'underline', 'subscript', 'superscript'
    ],
  };

public editorsReady(editor) {
      editor.ui.getEditableElement().parentElement.insertBefore(
        editor.ui.view.toolbar.element,
        editor.ui.getEditableElement()
      );
  }
```

**component.html**
```<ckeditor (ready)="editorsReady($event)" [editor]="Editor" [formControl]="editNewsEvent.controls['corpoNews']" id="corpoNews" name="corpoNews" [config]="optionsNews"></ckeditor>```

**ckeditor.js**
```
/**
 * @license Copyright (c) 2003-2019, CKSource - Frederico Knabben. All rights reserved.
 * For licensing, see LICENSE.md or https://ckeditor.com/legal/ckeditor-oss-license
 */

// The editor creator to use.
import ClassicEditorBase from '@ckeditor/ckeditor5-editor-classic/src/classiceditor';

import Essentials from '@ckeditor/ckeditor5-essentials/src/essentials';
import UploadAdapter from '@ckeditor/ckeditor5-adapter-ckfinder/src/uploadadapter';
import Autoformat from '@ckeditor/ckeditor5-autoformat/src/autoformat';
import Bold from '@ckeditor/ckeditor5-basic-styles/src/bold';
import Italic from '@ckeditor/ckeditor5-basic-styles/src/italic';
import BlockQuote from '@ckeditor/ckeditor5-block-quote/src/blockquote';
import CKFinder from '@ckeditor/ckeditor5-ckfinder/src/ckfinder';
import EasyImage from '@ckeditor/ckeditor5-easy-image/src/easyimage';
import Heading from '@ckeditor/ckeditor5-heading/src/heading';
import Image from '@ckeditor/ckeditor5-image/src/image';
import ImageCaption from '@ckeditor/ckeditor5-image/src/imagecaption';
import ImageStyle from '@ckeditor/ckeditor5-image/src/imagestyle';
import ImageToolbar from '@ckeditor/ckeditor5-image/src/imagetoolbar';
import ImageUpload from '@ckeditor/ckeditor5-image/src/imageupload';
import Indent from '@ckeditor/ckeditor5-indent/src/indent';
import Link from '@ckeditor/ckeditor5-link/src/link';
import List from '@ckeditor/ckeditor5-list/src/list';
import MediaEmbed from '@ckeditor/ckeditor5-media-embed/src/mediaembed';
import Paragraph from '@ckeditor/ckeditor5-paragraph/src/paragraph';
import PasteFromOffice from '@ckeditor/ckeditor5-paste-from-office/src/pastefromoffice';
import Table from '@ckeditor/ckeditor5-table/src/table';
import TableToolbar from '@ckeditor/ckeditor5-table/src/tabletoolbar';

import Underline from '@ckeditor/ckeditor5-basic-styles/src/underline'; //added underline plugin

export default class ClassicEditor extends ClassicEditorBase {}

// Plugins to include in the build.
ClassicEditor.builtinPlugins = [
	Essentials,
	UploadAdapter,
	Autoformat,
	Bold,
	Italic,
	BlockQuote,
	CKFinder,
	EasyImage,
	Heading,
	Image,
	ImageCaption,
	ImageStyle,
	ImageToolbar,
	ImageUpload,
	Indent,
	Link,
	List,
	MediaEmbed,
	Paragraph,
	PasteFromOffice,
	Table,
	TableToolbar,
	Underline  //added underline plugin
];

// Editor configuration.
ClassicEditor.defaultConfig = {
	toolbar: {
		items: [
			'heading',
			'|',
			'bold',
			'italic',
			'link',
			'bulletedList',
			'numberedList',
			'|',
			'indent',
			'outdent',
			'|',
			'imageUpload',
			'blockQuote',
			'insertTable',
			'mediaEmbed',
			'undo',
			'redo',
			'underline' // added underline plugin
		]
	},
	image: {
		toolbar: [
			'imageStyle:full',
			'imageStyle:side',
			'|',
			'imageTextAlternative'
		]
	},
	table: {
		contentToolbar: [
			'tableColumn',
			'tableRow',
			'mergeTableCells'
		]
	},
	// This value must be kept in sync with the language defined in webpack.config.js.
	language: 'en'
};
```

**Comment by ma2ciek** [comment]
Did you build the `ckeditor5-build-classic` with the added plugin after cloning it like it's in the documentation? 

I see that you have:
```js
import * as ClassicEditor from '@ckeditor/ckeditor5-build-classic';
```

So you load the original package instead of loading the file with the result of building the changed `ckeditor5-build-classic` package. You should have something like this pointing your custom build instead:

```js
import * as ClassicEditor from './path/to/ckeditor';
```

**Comment by Eyegabry99** [comment]
I don't have a custom build. I just edited the preset build

**Comment by ma2ciek** [comment]
So you need to continue the https://ckeditor.com/docs/ckeditor5/latest/builds/guides/integration/installing-plugins.html and generate the build.

**Comment by Eyegabry99** [comment]
> So you need to continue the https://ckeditor.com/docs/ckeditor5/latest/builds/guides/integration/installing-plugins.html and generate the build.

How can i continue? I'm stucked to yarn build. After that the guide is end, isn't it?

**Comment by Eyegabry99** [comment]
How can i continue? Isn't the guide ending with "Yarn build"?

**Comment by ma2ciek** [comment]
So after the `yarn build` command you should have generated build in the `build/ckeditor.js` file. That file you need to copy to the Angular project and import from the component.

**Comment by Eyegabry99** [comment]
What if i use npm? I launched npm install. Is this the correct command to build that?

**Comment by ma2ciek** [comment]
Nope. It's `npm run build`.

**Comment by Eyegabry99** [comment]
I run the build into folder node_modules > @ckeditor > ckeditor-build-classics. Is this the correct way?

**Comment by ma2ciek** [comment]
> I did that command into the following directory: node_modules > @ckeditor > ckeditor-build-classics. Is this the correct way? 

No, you should download the `ckeditor5-build-classic` package outside of the node_modules directory (like a new project). Otherwise, you will lose all your changes done inside `node_modules` after calling `npm install`. Then after installing and generating the custom build, you should copy the file to the angular project and import it from the component.

**Comment by getrealtec** [comment]
Ok CKEditor, this is way too complicated. The goal of libraries and frameworks is to simplify our lives. Do you realize that to add a simple : "underline" button you have to work for 30 minutes and reading tons of documentation and go completely out of standard ES6 principles?

What happened to :

import {Underline} from ckeditor;
use Underline where it counts...

Yes you are fully configurable and I've always loved CKEditor in the past, but this system you are suggesting is literally ridiculous. CKEditor needs to simplify how its customers are using its library. Adding a new button should be as simple as :
use NEW_BUTTON 
or 
config={toolbar: ["underline"]} 
where the config merges with the previous one. Maybe create a 
extraConfig={...}
that will merge current configs with the one added in extraConfig if you don't wish to automatically merge all configs put in "config".

Up your game please.

**Comment by ma2ciek** [comment]
Hey @getrealtec!

The problem is that we currently can't use plain imports to CKEditor 5 source files inside the Angular application. That works easily outside of it. The reason for this are conflicting loaders of Angular and CKEditor 5 and from Angular@5 there's no easy way to force Angular to use different loaders for some source files. So as the `Underline` plugin imports CSS and SVG files there's no easy way to import it inside of Angular application. I know it's a PITA, but after the Angular stopped exposing `ng eject` we have really no easy way to build CKEditor 5 sources together with Angular ones. Though, some work has been made in this topic - #26 - this is an unsupported version still, but you might wanna try it.

For easier setup you can try the [CKEditor 5 online builder](https://ckeditor.com/ckeditor-5/online-builder/), just pick plugins you want to use, download ready to use build and attach it inside your Angular application.

**Comment by martinsotirov** [comment]
Half a year later it's still extremely difficult to extend the classic build with plugins in an Angular environment. The effort is just not worth it for the few things most people need to extend it with (underline, alignment etc.).

**Comment by ma2ciek** [comment]
[Building editor from the source](https://github.com/ckeditor/ckeditor5-angular/issues/26) could make the integration much easier. Unfortunately, currently, I don't have time to look at it deeper so if anyone is willing to help feel welcome to investigate this topic.

**Comment by sarrasarra1** [comment]
hi! 
you can use =>  npm install --save ckeditor5-build-classic-plus 
 import ClassicEditor from 'ckeditor5-build-classic-plus'; 
  and problem resolved for underline :) 👍

**Comment by nadavsinai** [comment]
Writing here with hope someone finds this useful.
if you want to use CKEditor5 with Angular (12 in my case) you can make it work using @angular-builders/custom-webpack of ngx-build-plus  - the choice is yours and the concept is the same - you have to change the builder or your app to that custom builder and supply a webpack.confg.js file (just like they state in the docs for those 2 helper libs) in the webpack.config you have to do 2 things:
1. add the custom loaders for the postcss transform ckeditor needs (and svg) like this
```
{
            test: /ckeditor5-[^/\\]+[/\\]theme[/\\]icons[/\\][^/\\]+\.svg$/,
            use: ['raw-loader']
        },
        {
            test: /ckeditor5-[^/\\]+[/\\]theme[/\\].+\.css$/,
            use: [
                {
                    loader: 'style-loader',
                    options: {
                        injectType: 'singletonStyleTag',
                        attributes: {
                            'data-cke': true
                        }
                    }
                },
                'css-loader',
                {
                    loader: 'postcss-loader',
                    options: {
                        postcssOptions: styles.getPostCssConfig({
                            themeImporter: {
                                themePath: require.resolve('@ckeditor/ckeditor5-theme-lark')
                            },
                            minify: false
                        })
                    }
                }
            ]
        },
    )
    ```
2.    change the default logic for angular CSS loading rules to exclude the ckeditor css
like this
``` const cssRule = config.module.rules.find(x => x.test.exec && x.test.exec('.css'));
    cssRule.exclude = /ckeditor5/;`
 ```
  to do that - you need to make your webpack.config.js return a function rather than an object
  
  I have a working project here  [angular-ckeditor](https://github.com/nadavsinai/angular-ckeditor/)
happy hacking ;-)

**Comment by steve-tapley** [comment]
Adding on to the comments from @nadavsinai & @getrealtec 

The solution proposed by @nadavsinai works for us, but it took us a long time to find this solution.  

The angular documentation in this regard seems particularly poor or just plain missing.  For custom builds, we had to cobble together a working configuration from the Angular, React & Vue docs which wasn't a great developer experience.

I understand the configuration proposed by @nadavsinai is not recommended, but its the only way for us to actually use CK5 in our angular application.

Could you re-open this issue and just improve your documentation?

**Comment by nakhai9** [comment]
Presently, I'm using Angular 17 and Micro Frontend for my app. How do I embed the underline plugin in my project?

I have tried to implement 'underline' to config.

import { Component } from '@angular/core';
import { CommonModule } from '@angular/common';
import { CKEditorModule } from '@ckeditor/ckeditor5-angular';
import * as ClassicEditor from '@ckeditor/ckeditor5-build-classic';
import { ChangeEvent, CKEditorComponent } from '@ckeditor/ckeditor5-angular/ckeditor.component';

@Component({
  selector: 'web-content',
  standalone: true,
  imports: [CommonModule, CKEditorModule],
  templateUrl: './web-content.component.html',
  styleUrl: './web-content.component.scss'
})
export class WebContentComponent {
  title: string = 'Web Content';
  description: string = `Lorem ipsum dolor sit amet consectetur. Aliquet sit odio aenean proin mi. Enim aliquam auctor
  elit elementum. Blandit lectus mattis lacus tincidunt. Tortor volutpat nec integer
  sollicitudin massa aliquam euismod.`;

  Editor = ClassicEditor;

  data: string = `Hello MFE`;

  config = {
    toolbar: {
      items: ['bold', 'italic', 'underline' , '|', 'bulletedList', 'numberedList', '|', 'link']
    }
  };

  onReady(editor: CKEditorComponent) {
    console.log('CKEditor5 Angular Component is ready to use!', editor);
  }

  onChange({ editor }: ChangeEvent) {
    this.data = editor.data.get();
  }
}
